From 40946a988815952c81185db06f998b4f07a15fd3 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Thu, 20 Oct 2022 20:05:56 +0800
Subject: [PATCH 031/595] [Huawei] cgroup: Fix kabi broken in mem_cgroup and
 blkcg

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187773

-------------------------------

In struct mem_cgroup, we add two members: struct cgroup_subsys_state
*wb_blk_css and struct list_add memcg_node; the size of
struct cgroup_subsys_state* is 8 bytes. the size of struct list_head
is 16 bytes. Therefore, it should use 3 KABI_RESERVE.
In struct blkcg, we add one member: struct list_add memcg_list. The size
of struct list_head is 16 bytes. Therefore, it should use 2 KABI_RESERVE

Fixes: 13b8199f0a47 ("cgroup: support cgroup writeback on cgroupv1")
Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/blk-cgroup.h | 2 +-
 include/linux/memcontrol.h | 9 ++++++---
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/include/linux/blk-cgroup.h b/include/linux/blk-cgroup.h
index 4c1875d69c36..86a10c5acf9b 100644
--- a/include/linux/blk-cgroup.h
+++ b/include/linux/blk-cgroup.h
@@ -63,8 +63,8 @@ struct blkcg {
 	struct list_head		memcg_list;
 #else
 	KABI_RESERVE(1)
-#endif
 	KABI_RESERVE(2)
+#endif
 	KABI_RESERVE(3)
 	KABI_RESERVE(4)
 };
diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index a6483f69768f..27beafcb5fb6 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -375,14 +375,17 @@ struct mem_cgroup {
 #ifdef CONFIG_DYNAMIC_HUGETLB
 	struct dhugetlb_pool *hpool;
 #endif
+#ifdef CONFIG_CGROUP_WRITEBACK
+	KABI_USE(1, struct cgroup_subsys_state *wb_blk_css)
+#else
+	KABI_RESERVE(1)
+#endif
 #if defined(CONFIG_CGROUP_WRITEBACK) && !defined(__GENKSYMS__)
-	struct cgroup_subsys_state *wb_blk_css;
 	struct list_head memcg_node;
 #else
-	KABI_RESERVE(1)
 	KABI_RESERVE(2)
-#endif
 	KABI_RESERVE(3)
+#endif
 	KABI_RESERVE(4)
 	KABI_RESERVE(5)
 	KABI_RESERVE(6)
-- 
2.17.1

